packages/flutter/lib/src/material/slider_theme.dart

```diff
@@ -638,8 +617,42 @@ class PaddleSliderValueIndicatorShape extends SliderComponentShape {
     return _bottomLobeEnd;
   }
 
-  void _drawValueIndicator(Canvas canvas, Offset center, Paint paint, double scale,
-      TextPainter labelPainter, double textScaleFactor) {
+  // Determines the "best" offset to keep the bubble on the screen. The calling
+  // code will bound that with the available movement in the paddle shape.
+  double _getIdealOffset(RenderBox parentBox, double halfWidthNeeded, double scale, Offset center) {
+    const double edgeMargin = 4.0;
+    final Rect topLobeRect = new Rect.fromLTWH(
+      -_topLobeRadius - halfWidthNeeded,
+      -_topLobeRadius - _distanceBetweenTopBottomCenters,
+      2.0 * (_topLobeRadius + halfWidthNeeded),
+      2.0 * _topLobeRadius,
+    );
+    final Offset origin = center - parentBox.localToGlobal(Offset.zero);
+    // We can just multiply by scale instead of a transform, since we're scaling
+    // around (0, 0).
+    final Offset topLeft = parentBox.localToGlobal((topLobeRect.topLeft * scale) + origin);
+    final Offset bottomRight = parentBox.localToGlobal((topLobeRect.bottomRight * scale) + origin);
```

Comment:
localToGlobal can be expensive; could this be computed by scaling the parent box's width/height and adding them to topLeft?

Reply:
But what if the parent box is rotated? What I really want is to ask the canvas what the matrix on the top of the stack is, but we don't have that ability.

Reply:
Good point, I guess we always have assume that a widget might be transformed relative to its parent.